File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/infrastructure/persistence/adapters/RefreshTokenRepositoryAdapter.java
package br.com.deltaglobalbank.identity.infrastructure.persistence.adapters;

import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.identity.domain.token.RefreshToken;
import br.com.deltaglobalbank.identity.domain.token.RefreshTokenRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.RefreshTokenEntity;
import br.com.deltaglobalbank.identity.infrastructure.persistence.mappers.RefreshTokenMapper;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaRefreshTokenRepository;
import org.springframework.stereotype.Component;

@Component
public class RefreshTokenRepositoryAdapter implements RefreshTokenRepository {

    private final JpaRefreshTokenRepository jpaRefreshTokenRepository;

    public RefreshTokenRepositoryAdapter(JpaRefreshTokenRepository jpaRefreshTokenRepository) {
        this.jpaRefreshTokenRepository = jpaRefreshTokenRepository;
    }

    @Override
    public RefreshToken findById(UUID id) {
        return jpaRefreshTokenRepository.findById(id).map(RefreshTokenMapper::toDomain).orElse(null);
    }

    @Override
    public RefreshToken findByTokenHash(String hash) {
        RefreshTokenEntity entity = jpaRefreshTokenRepository.findByTokenHash(hash);
        return entity != null ? RefreshTokenMapper.toDomain(entity) : null;
    }

    @Override
    public List<RefreshToken> findAllByUserId(UUID userId) {
        return jpaRefreshTokenRepository.findAllByUserId(userId).stream().map(RefreshTokenMapper::toDomain).toList();
    }

    @Override
    public List<RefreshToken> findAllActiveByUserId(UUID userId) {
        return jpaRefreshTokenRepository.findAllByUserIdAndRevokedAtIsNull(userId).stream()
            .map(RefreshTokenMapper::toDomain)
            .toList();
    }

    @Override
    public RefreshToken save(RefreshToken refreshToken) {
        RefreshTokenEntity existing = jpaRefreshTokenRepository.findById(refreshToken.getId()).orElse(null);
        RefreshTokenEntity entityToSave = existing != null
            ? RefreshTokenMapper.applyTo(refreshToken, existing)
            : RefreshTokenMapper.toEntity(refreshToken);
        return RefreshTokenMapper.toDomain(jpaRefreshTokenRepository.save(entityToSave));
    }

    @Override
    public List<RefreshToken> saveAll(List<RefreshToken> refreshTokens) {
        List<RefreshTokenEntity> entities = refreshTokens.stream()
            .map(token -> {
                RefreshTokenEntity existing = jpaRefreshTokenRepository.findById(token.getId()).orElse(null);
                return existing != null
                    ? RefreshTokenMapper.applyTo(token, existing)
                    : RefreshTokenMapper.toEntity(token);
            })
            .toList();
        return jpaRefreshTokenRepository.saveAll(entities).stream().map(RefreshTokenMapper::toDomain).toList();
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/infrastructure/persistence/adapters/RefreshTokenRepositoryAdapter.kt
package br.com.deltaglobalbank.identity.infrastructure.persistence.adapters

import br.com.deltaglobalbank.identity.domain.token.RefreshToken
import br.com.deltaglobalbank.identity.domain.token.RefreshTokenRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.mappers.applyTo
import br.com.deltaglobalbank.identity.infrastructure.persistence.mappers.toDomain
import br.com.deltaglobalbank.identity.infrastructure.persistence.mappers.toEntity
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaRefreshTokenRepository
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class RefreshTokenRepositoryAdapter(
    private val jpaRefreshTokenRepository: JpaRefreshTokenRepository
) : RefreshTokenRepository {

    override fun findById(id: UUID): RefreshToken? =
        jpaRefreshTokenRepository.findById(id).orElse(null)?.toDomain()

    override fun findByTokenHash(hash: String): RefreshToken? =
        jpaRefreshTokenRepository.findByTokenHash(hash)?.toDomain()

    override fun findAllByUserId(userId: UUID): List<RefreshToken> =
        jpaRefreshTokenRepository.findAllByUserId(userId).map { it.toDomain() }

    override fun findAllActiveByUserId(userId: UUID): List<RefreshToken> =
        jpaRefreshTokenRepository.findAllByUserIdAndRevokedAtIsNull(userId).map { it.toDomain() }

    override fun save(refreshToken: RefreshToken): RefreshToken {
        val existing = jpaRefreshTokenRepository.findById(refreshToken.id).orElse(null)
        val entityToSave = if (existing != null) refreshToken.applyTo(existing) else refreshToken.toEntity()
        return jpaRefreshTokenRepository.save(entityToSave).toDomain()
    }

    override fun saveAll(refreshTokens: List<RefreshToken>): List<RefreshToken> {
        val entities = refreshTokens.map { token ->
            val existing = jpaRefreshTokenRepository.findById(token.id).orElse(null)
            if (existing != null) token.applyTo(existing) else token.toEntity()
        }
        return jpaRefreshTokenRepository.saveAll(entities).map { it.toDomain() }
    }
}
